### ISSUE TITLE
Package shuffling

### ISSUE THREAD

**Issue by dhirschfeld** [conda/conda#6284]
Recently all my conda envs have wanted to shuffle packages between the defaults and conda-forge channels. The packages affected are shown in the output below.

Taking `bzip2` as an example, conda wants to change from the defaults `1.0.6=vc14hdec8e7a_1`  to the conda-forge `1.0.6=vc14_1` version.  To me, because they both have the same version number they should be considered equivalent and the tie should be broken by the channel ordering. Because defaults is higher than conda-forge I would expect that the defaults version should win and there's be no shuffling of packages.
```
Package plan for installation in environment C:\Anaconda3:

The following NEW packages will be INSTALLED:

    backports.functools_lru_cache: 1.4-py36_1                 conda-forge
    krb5:                          1.14.2-vc14_0              conda-forge [vc14]

The following packages will be UPDATED:

    bzip2:                         1.0.6-vc14hdec8e7a_1                   --> 1.0.6-vc14_1       conda-forge [vc14]
    gensim:                        3.0.1-py36_vc14hb0e632d_0              --> 3.1.0-py36_vc14_0  conda-forge [vc14]
    hdf5:                          1.10.1-vc14hb361328_0                  --> 1.10.1-vc14_1      conda-forge [vc14]
    jpeg:                          9b-vc14h4d7706e_1                      --> 9b-vc14_2          conda-forge [vc14]
    libssh2:                       1.8.0-vc14hcf584a9_2                   --> 1.8.0-vc14_2       conda-forge [vc14]
    libxslt:                       1.1.29-vc14hf85b8d4_5                  --> 1.1.29-vc14_5      conda-forge [vc14]
    matplotlib:                    2.1.0-py36h11b4b9c_0                   --> 2.1.0-py36_1       conda-forge
    pillow:                        4.2.1-py36hdb25ab2_0                   --> 4.3.0-py36_1       conda-forge
    snappy:                        1.1.7-vc14h2dea872_1                   --> 1.1.7-vc14_1       conda-forge [vc14]

The following packages will be DOWNGRADED:

    boost:                         1.65.1-py36_vc14h50d208f_3             --> 1.65.1-py36_vc14_0 conda-forge [vc14]
    boost-cpp:                     1.65.1-vc14h7942e50_3                  --> 1.65.1-vc14_0      conda-forge [vc14]
    curl:                          7.55.1-vc14hdaba4a4_3                  --> 7.55.1-vc14_0      conda-forge [vc14]
    freetype:                      2.8-vc14h17c9bdf_0                     --> 2.7-vc14_2         conda-forge [vc14]
    icu:                           58.2-vc14hc45fdbb_0                    --> 58.1-vc14_1        conda-forge [vc14]
    libiconv:                      1.15-vc14h29686d3_5                    --> 1.14-vc14_4        conda-forge [vc14]
    libpng:                        1.6.32-vc14h5163883_3                  --> 1.6.28-vc14_2      conda-forge [vc14]
    libtiff:                       4.0.8-vc14h04e2a1e_10                  --> 4.0.7-vc14_1       conda-forge [vc14]
    lzo:                           2.10-vc14h0a64fa6_1                    --> 2.10-vc14_0        conda-forge [vc14]
    openssl:                       1.0.2l-vc14hcac20b0_2                  --> 1.0.2l-vc14_0      conda-forge [vc14]
    qt:                            5.6.2-vc14h6f8c307_12                  --> 5.6.2-vc14_1       conda-forge [vc14]
    sqlite:                        3.20.1-vc14h7ce8c62_1                  --> 3.19.3-vc14_1      conda-forge [vc14]
    tk:                            8.6.7-vc14hb68737d_1                   --> 8.6.6-vc14_5       conda-forge [vc14]
    yaml:                          0.1.7-vc14hb31d195_1                   --> 0.1.6-vc14_0       conda-forge [vc14]
    zlib:                          1.2.11-vc14h1cdd9ab_1                  --> 1.2.11-vc14_0      conda-forge [vc14]
```

**`.condarc`**
```
channel_priority: false
channels:
  - defaults
  - dhirschfeld
  - conda-forge
  - numba
```

I'm wondering if this is another consequence of timestamp prioritization? c.f. #6271

**Comment by dhirschfeld** [comment]
If it's worthwhile debugging the verbose output log is avaialble at:
https://gist.github.com/dhirschfeld/e3e0427a7d226c88ae63a9d3e305affd

**Comment by kalefranz** [comment]
@dhirschfeld What was the conda command you executed on your root environment to get the above output?  Guessing it was a `conda install`, but what specs to did you give?

**Comment by dhirschfeld** [comment]
I believew it was a `conda update conda`. I believe *any* update or install command will try to shuffle those packages though

**Comment by dhirschfeld** [comment]
***xref:*** #6281 seems to be the same  issue

**Comment by kalefranz** [comment]
For this shuffling issue, why have `channel_priority: false`?  Does flipping that to `true` cause behavior to be as expected?

**Comment by dhirschfeld** [comment]
I always have `channel_priority` to false because when I update I want the latest package irrespective of which channel that package comes from. I don't want the buildstring to factor in the calculation though.

IIUC channel priorities should only used to break ties however I'm seeing (still) packages flip back and forth between the conda-forge and defaults versions on almost any update/install command.

**Comment by dhirschfeld** [comment]
# MVCE:

OS: win64

**`.condarc`**
```
channels:
  - defaults
  - https://conda.anaconda.org/conda-forge

channel_priority: false
```
Note: `channel_priority` is false

**`shuffle.yml`**
```
name: shuffle

channels:
  - defaults

dependencies:
  - gflags
  - libpng
  - zlib <1.2.11
```
-----
Create the environment with an older `zlib`:
```
λ conda env create --file .\shuffle.yml
```

Try to update `zlib`:
```
λ activate shuffle
(shuffle) λ conda update zlib

Solving environment: done

## Package Plan ##

  environment location: C:\Miniconda3\envs\shuffle

  added / updated specs:
    - zlib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    zlib-1.2.11                |       h8395fce_2         237 KB

The following packages will be UPDATED:

    gflags: 2.2.1-h6538335_0 --> 2.2.1-vc14_0      conda-forge [vc14]  <-- !!!!!!!!!!!!!!!!!!!!!!
    libpng: 1.6.30-vc14_1    --> 1.6.34-vc14_0     conda-forge [vc14]  <-- !!!!!!!!!!!!!!!!!!!!!!
    zlib:   1.2.8-vc14_3     --> 1.2.11-h8395fce_2

Proceed ([y]/n)? y
```

Note the highlighted lines - this is what I believe to be incorrect behaviour (a bug)

The update of a completely unrelated package (`zlib`) has triggered a change in `gflags`.

When I have `channel_priority=false` I only want to update an existing package when there is a greater version available - the channel should play no part in selecting which package to install. Only when installing a *new* package should the `channel_priority` ever be used to break ties.

In this case I have version `2.2.1` of `gflags` installed from the `defaults` channel and the solver wants to replace it with the same version `2.2.1` but from the `conda-forge` channel which is incorrect according to my `channel_priority` setting.

Even if I did have  `channel_priority=true`, `conda-forge` has a lower priority than `defaults` so it still shouldn't choose to replace a package with the same version from a lower priority channel.

After accepting the above changes I now have the new `zlib` *and* the *"UPDATED"* `gflags` and `libpng` so that should be the latest of everything, however if I run an `update --all` conda will choose to revert the changes made to `gflags` and `libpng`:
```
(shuffle) λ conda update --all
Solving environment: done

## Package Plan ##

  environment location: C:\Miniconda3\envs\shuffle


The following packages will be UPDATED:

    gflags: 2.2.1-vc14_0  conda-forge [vc14] --> 2.2.1-h6538335_0
    libpng: 1.6.34-vc14_0 conda-forge [vc14] --> 1.6.34-h79bbb47_0

Proceed ([y]/n)? y
```

On every single run of conda a whole bunch of packages will be *"UPDATED"* to the other channel causing every change to take a lot more time than it would otherwise.

**Comment by dhirschfeld** [comment]
If I change `channel_priority` to `true` and repeat it looks like it wants to shuffle `gflags` to the lower priority channel but doesn't touch `libpng` in this case:
```
(shuffle) λ conda update zlib
Solving environment: done

## Package Plan ##

  environment location: C:\Miniconda3\envs\shuffle

  added / updated specs:
    - zlib


The following packages will be UPDATED:

    gflags: 2.2.1-h6538335_0 --> 2.2.1-vc14_0      conda-forge [vc14]
    zlib:   1.2.8-vc14_3     --> 1.2.11-h8395fce_2

Proceed ([y]/n)?
```

**Comment by dhirschfeld** [comment]
Hopefully this helps track it down...

It's a somewhat benign bug, but in practice is very annoying!

**Comment by kalefranz** [comment]
Can you confirm conda version with ‘conda info’?

**Comment by dhirschfeld** [comment]
```
     active environment : None
       user config file : C:\Users\dhirschf\.condarc
 populated config files : C:\Users\dhirschf\.condarc
          conda version : 4.4.10
    conda-build version : 3.4.2
         python version : 3.6.4.final.0
       base environment : C:\Miniconda3  (writable)
           channel URLs : https://repo.continuum.io/pkgs/main/win-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/win-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/win-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/win-64
                          https://repo.continuum.io/pkgs/pro/noarch
                          https://repo.continuum.io/pkgs/msys2/win-64
                          https://repo.continuum.io/pkgs/msys2/noarch
                          https://conda.anaconda.org/conda-forge/win-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : C:\Miniconda3\pkgs
                          C:\Users\dhirschf\AppData\Local\conda\conda\pkgs
       envs directories : C:\Miniconda3\envs
                          C:\Users\dhirschf\AppData\Local\conda\conda\envs
                          C:\Users\dhirschf\.conda\envs
               platform : win-64
             user-agent : conda/4.4.10 requests/2.18.4 CPython/3.6.4 Windows/7 Windows/6.1.7601
          administrator : True
             netrc file : None
           offline mode : False
```

**Comment by mbargull** [comment]
The explanation for this behavior is that there is
1. different assignment of `features` in `conda-forge` compared to `defaults`, and
2. feature prioritization is handled differently for explicitly requested and dependency packages.

Regarding the first point, I assume @mcg1969 would say that, e.g., `fn: 'libpng-1.6.34-vc14_0', features: 'vc14', depends: ['vc 14', 'zlib 1.2.11']` on the `conda-forge` package is a wrong feature/dependency assignment...
The second point is just status quo and sadly leads to inconsistent behavior between installing a package explicitly or only as a dependency. (As an example for this, you can compare the outputs of `conda install -c conda-forge numpy` and `conda install -c conda-forge xnumpy`.)

----
The following are just the explanations I assume for the observed behavior:

----
Package information (URL, features, dependencies)
```bash
$ alias parse_info='python -c '"'"'import sys,json; print(*(" ".join(map(repr, (x["url"], x.get("features"), x["depends"]))) for x in sum(json.load(sys.stdin).values(), [])), sep="\n")'"'"
$ conda info --json gflags=2.2.1=h6538335_0 gflags=2.2.1=vc14_0 libpng=1.6.30=vc14_1 libpng=1.6.34=h79bbb47_0 libpng=1.6.34=vc14_0 zlib=1.2.8=vc14_3 zlib=1.2.11=h8395fce_2 vc=14=h0510ff6_3 | parse_info
'https://repo.continuum.io/pkgs/main/win-64/gflags-2.2.1-h6538335_0.tar.bz2' None ['vc 14.*']
'https://conda.anaconda.org/conda-forge/win-64/gflags-2.2.1-vc14_0.tar.bz2' 'vc14' []
'https://repo.continuum.io/pkgs/free/win-64/libpng-1.6.30-vc14_1.tar.bz2' None ['vc 14', 'zlib 1.2.*']
'https://repo.continuum.io/pkgs/main/win-64/libpng-1.6.34-h79bbb47_0.tar.bz2' None ['vc 14.*', 'zlib >=1.2.11,<1.3.0a0']
'https://conda.anaconda.org/conda-forge/win-64/libpng-1.6.34-vc14_0.tar.bz2' 'vc14' ['vc 14', 'zlib 1.2.11']
'https://repo.continuum.io/pkgs/main/win-64/vc-14-h0510ff6_3.tar.bz2' None ['vs2015_runtime >=14.0.25123,<15.0a0']
'https://repo.continuum.io/pkgs/main/win-64/zlib-1.2.11-h8395fce_2.tar.bz2' None ['vc 14.*']
'https://repo.continuum.io/pkgs/free/win-64/zlib-1.2.8-vc14_3.tar.bz2' None ['vc 14']
'https://conda.anaconda.org/conda-forge/win-64/zlib-1.2.8-vc14_3.tar.bz2' 'vc14' []
```

----
* maximize `(version, channel)` of requested packages (`zlib`):
  picks `zlib` from `defaults`
* maximize packages with features:
  picks `libpng`, `gflags` from `conda-forge` as those use the `vc14` feature
* maximize `(version, channel)` of dependencies
```
(shuffle) λ conda update zlib
...
    gflags: 2.2.1-h6538335_0 --> 2.2.1-vc14_0      conda-forge [vc14]  <-- !!!!!!!!!!!!!!!!!!!!!!
    libpng: 1.6.30-vc14_1    --> 1.6.34-vc14_0     conda-forge [vc14]  <-- !!!!!!!!!!!!!!!!!!!!!!
    zlib:   1.2.8-vc14_3     --> 1.2.11-h8395fce_2
```

----
* maximize `(version, channel)` of requested packages (`zlib`, `libpng`, `gflags`, ...):
  picks up all packages from `defaults` since featured packages are not maximized beforehand 

Versions tied for packages from both channels, hence the ones from higher priority channel are used. (This has to be done to avoid comparing build numbers across channels.)
```
(shuffle) λ conda update --all
...
    gflags: 2.2.1-vc14_0  conda-forge [vc14] --> 2.2.1-h6538335_0
    libpng: 1.6.34-vc14_0 conda-forge [vc14] --> 1.6.34-h79bbb47_0
```

----
> If I change channel_priority to true and repeat it looks like it wants to shuffle gflags to the lower priority channel but doesn't touch libpng in this case:

(I'm not sure yet what happened here...)
```
(shuffle) λ conda update zlib
....
    gflags: 2.2.1-h6538335_0 --> 2.2.1-vc14_0      conda-forge [vc14]
    zlib:   1.2.8-vc14_3     --> 1.2.11-h8395fce_2
```

**Comment by mbargull** [comment]
Well the last case I couldn't explain, I also can't reproduce. @dhirschfeld can you give the exact sequence of commands, starting with `conda create` for a new environment, which reproduce this (meaning the `gflags` but not `libpng` update)?

**Comment by dhirschfeld** [comment]
There's a fully reproducible test case in the example I posted above. Create the initial environment with the `shuffle.yml` file posted and follow the commands

**Comment by dhirschfeld** [comment]
> *Versions tied for packages from both channels, hence the ones from higher priority channel are used*

...except the packages chosen are from the `conda-forge` channel which has a *lower* priority!

**Comment by mbargull** [comment]
> ...except the packages chosen are from the `conda-forge` channel which has a lower priority!

I tried to be *very precise* in my post above by using the horizontal delimiters to separate the *different cases* you described. But to clarify again, my comment was only meant for this part:

----
* maximize `(version, channel)` of requested packages (`zlib`, `libpng`, `gflags`, ...):
  picks up all packages from `defaults` since featured packages are not maximized beforehand 

Versions tied for packages from both channels, hence the ones from higher priority channel are used. (This has to be done to avoid comparing build numbers across channels.)
```
(shuffle) λ conda update --all
...
    gflags: 2.2.1-vc14_0  conda-forge [vc14] --> 2.2.1-h6538335_0
    libpng: 1.6.34-vc14_0 conda-forge [vc14] --> 1.6.34-h79bbb47_0
```

----
^-- notice that (for *only this part* of your problem description) we have exactly the case you want, the packages from the lower priority channel get replaced by the ones from the higher priority one.

**Comment by mbargull** [comment]
> There's a fully reproducible test case in the example I posted above. Create the initial environment with the `shuffle.yml` file posted and follow the commands

You wrote before
> If I change `channel_priority` to `true` and repeat it looks like it wants to shuffle `gflags` to the lower priority channel but doesn't touch `libpng` in this case:

From which I could not deduce what "and repeat" encompasses. By only repeating the environment creation followed by `conda update zlib` I could not reproduce what you saw. Hence I was asking for the complete reproducer for this last case.

**Comment by dhirschfeld** [comment]
@mbargull - you're right sorry, your comment was referring to the `update all` command rather than the `update zlib` command which initially chose to change the packages for the same version from the lower priority `conda-forge` channel.

The main problem I think is the initial `update zlib` install plan which (as mentioned above) changes the unrelated `libpng` and `gflags` packages for those from a lower priority channel.

The `update all` example after that was mainly to demonstrate the inconsistency in the original `update zlib` plan.

From your explanation it sounds like it might be a nasty interaction with features from different channels? If that's not something the conda solver can fix perhaps someone could suggest how to fix all the `conda-forge` packages so as to not trigger this package shuffling behaviour?

I'm afraid I don't have enough of an understanding of how to do that myself but if someone explains what is wrong with the `conda-forge` recipes and how to fix them I'm happy to open PRs against all mis-specified recipes.

As it is the end-user experience is pretty painful so it would be great to find some way forward with this!

**Comment by dhirschfeld** [comment]
Re the `channel_priority=true` test *"and repeat"* just meant nuke the shuffle environment and repeat all the same steps as for the `channel_priority=false` - e.g.
```
conda env create --file .\shuffle.yml
activate shuffle
conda update zlib
```
...however because I didn't nuke my environment properly my results above weren't correct.

It turns out that the install plan for `update zlib` is exactly the same irrespective of whether `channel_priority` is true or false:
```
λ(shuffle)  conda update zlib
Solving environment: done

## Package Plan ##

  environment location: C:\Miniconda3\envs\shuffle

  added / updated specs:
    - zlib


The following packages will be UPDATED:

    gflags: 2.2.1-h6538335_0 --> 2.2.1-vc14_0      conda-forge [vc14]
    libpng: 1.6.30-vc14_1    --> 1.6.34-vc14_0     conda-forge [vc14]
    zlib:   1.2.8-vc14_3     --> 1.2.11-h8395fce_2

Proceed ([y]/n)? n
```
```
λ(shuffle)  cat C:\Users\dhirschf\.condarc
channels:
  - defaults
  - https://conda.anaconda.org/conda-forge

channel_priority: true
```

**Comment by dhirschfeld** [comment]
Whilst the main problem is the shuffling of gflags/libpng with the `update zlib` command (possibly due to features) the `update all` also demonstrates undesirable behaviour (IMHO) which I'll detail below.

```
(shuffle) λ conda update --all
...
    gflags: 2.2.1-vc14_0  conda-forge [vc14] --> 2.2.1-h6538335_0
    libpng: 1.6.34-vc14_0 conda-forge [vc14] --> 1.6.34-h79bbb47_0
```
> *^-- notice that we have exactly the case you want, the packages from the lower priority channel get replaced by the ones from the higher priority one.*

This statement is incorrect - the above behaviour is exactly what I do not want!!

By specifying `channel_priority=false` I explicitly do not want the channel order to play any part in the selection of a package to install. The obvious question this raises is then how does the solver choose between two perfectly valid install plans - one with the `conda-forge` and the other with the `defaults` package?

My idea here is to apply a *"minimum shuffling"* constraint - i.e choose the plan which will minimise any changes to the existing environment. In this case that means that if one of the `defaults` or `conda-forge` packages is already installed and both are valid install plans then the plan with the already installed package will win. So in the specific `update all` example above, because the `conda-forge` packages are already installed the plan keeping the already installed packages (and making no changes) would win.

By specifying `channel_priority=false` I only want a package to change if there is a newer version available. Build strings, features or channel ordering shouldn't play a part (in so far as that is possible)

My thoughts on how to implement this are as briefly outlined in https://github.com/conda/conda/issues/6806#issuecomment-361151561 - i.e. include an `is_already_installed` flag so that the ordering would be:
```
(version_number, is_already_installed, channel_priority)
```
In this way `channel_priority` would only ever come into play when installing a new package such that the `is_already_installed` flag is False for all install plans.

**Comment by dhirschfeld** [comment]
Note: this issue is actually a dup of #6281 - they got there a day before me!

That issue demonstrates the same package shuffling behaviour but hopefully the MCVE here will assist in getting to the bottom of the problem.

**Comment by dhirschfeld** [comment]
As an end user what I would like is that any update or install command make only the minimum required changes to my existing environment.

Going back to the `update zlib` example, in this case updating one package (zlib), shouldn't make changes to any other unrelated packages e.g. gflags/libpng. I suspect that the packages aren't entirely unrelated because of the `vc14` feature, however because the end result after doing:
```
conda update zlib
conda update --all
```
...is that only zlib has changed from `1.2.8-vc14_3` to the newer `1.2.11-h8395fce_2` the original update plan for zlib should have chosen this final state and not made any unnecessary changes to either gflags or libpng (which were then reverted by the `update all`)

**Comment by mbargull** [comment]
>> ^-- notice that we have exactly the case you want, the packages from the lower priority channel get replaced by the ones from the higher priority one.
>
>This statement is incorrect - the above behaviour is exactly what I do not want!!

True, wrong choice of words, sorry.

**Comment by mbargull** [comment]
@dhirschfeld, the following is nothing that helps with your current situation, but in future this should improve: https://github.com/conda-forge/conda-forge.github.io/issues/537#issuecomment-368231456

**Comment by dhirschfeld** [comment]
Just adding some more evidence that this is a general usability problem - from fperez [on Gitter](https://gitter.im/jupyterlab/jupyterlab?at=5ad565196d7e07082b009d44):
> *using pip b/c my default env doesn't use conda-forge, and if I try to install jlab form c-forge, it tries to pull in a massive amount of packages.
it's really annoying that conda won't install just one package from another repo (and it's core deps at most)...
It's trying to update all my R packages, which have nothing to do with lab.
So I'll pull from pip for now.*

I see that `conda-smithy=3.0` has been released so hopefully that improves things but I guess all existing packages will have to be rebuild with cb3 before users will notice anything different

**Comment by shoyer** [comment]
Yes, this is a serious usability problem.

If I have NumPy installed from conda-forge, attempting to install *any* unrelated package from the default channel will insist on installing the default version of NumPy -- including MKL, which is a 143 MB download.

Example with NumPy and toolz (a pure Python dependency):

<details>

```
$ conda create --yes -n testing123 -c conda-forge numpy
Solving environment: done

## Package Plan ##

  environment location: /Users/shoyer/miniconda3/envs/testing123

  added / updated specs:
    - numpy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-1.14.4               |py36_blas_openblash24bf2e0_200         3.9 MB  conda-forge
    python-3.6.5               |                1        13.9 MB  conda-forge
    openblas-0.2.20            |                8        20.1 MB  conda-forge
    certifi-2018.4.16          |           py36_0         142 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        38.1 MB

The following NEW packages will be INSTALLED:

    blas:            1.1-openblas                          conda-forge
    ca-certificates: 2018.4.16-0                           conda-forge
    certifi:         2018.4.16-py36_0                      conda-forge
    libgfortran:     3.0.0-0                               conda-forge
    ncurses:         5.9-10                                conda-forge
    numpy:           1.14.4-py36_blas_openblash24bf2e0_200 conda-forge [blas_openblas]
    openblas:        0.2.20-8                              conda-forge
    openssl:         1.0.2o-0                              conda-forge
    pip:             9.0.3-py36_0                          conda-forge
    python:          3.6.5-1                               conda-forge
    readline:        7.0-0                                 conda-forge
    setuptools:      39.2.0-py36_0                         conda-forge
    sqlite:          3.20.1-2                              conda-forge
    tk:              8.6.7-0                               conda-forge
    wheel:           0.31.0-py36_0                         conda-forge
    xz:              5.2.3-0                               conda-forge
    zlib:            1.2.11-h470a237_3                     conda-forge


Downloading and Extracting Packages
numpy-1.14.4         |  3.9 MB | : 100% 1.0/1 [00:05<00:00,  3.91s/it]
python-3.6.5         | 13.9 MB | : 100% 1.0/1 [00:07<00:00,  7.03s/it]
openblas-0.2.20      | 20.1 MB | : 100% 1.0/1 [00:08<00:00,  8.09s/it]
certifi-2018.4.16    |  142 KB | : 100% 1.0/1 [00:00<00:00,  1.67it/s]
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use:
# > source activate testing123
#
# To deactivate an active environment, use:
# > source deactivate
#

$ conda activate testing123
$ conda install toolz
Solving environment: done

## Package Plan ##

  environment location: /Users/shoyer/miniconda3/envs/testing123

  added / updated specs:
    - toolz


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-1.13.1               |           py36_0         3.6 MB
    mkl-2017.0.4               |       h1fae6ae_0       143.8 MB
    intel-openmp-2018.0.3      |                0        1004 KB
    certifi-2018.4.16          |           py36_0         142 KB
    ------------------------------------------------------------
                                           Total:       148.5 MB

The following NEW packages will be INSTALLED:

    intel-openmp:    2018.0.3-0
    mkl:             2017.0.4-h1fae6ae_0
    toolz:           0.9.0-py36_0

The following packages will be UPDATED:

    certifi:         2018.4.16-py36_0                      conda-forge --> 2018.4.16-py36_0
    openssl:         1.0.2o-0                              conda-forge --> 1.0.2o-h26aff7b_0

The following packages will be DOWNGRADED:

    blas:            1.1-openblas                          conda-forge --> 1.0-mkl
    ca-certificates: 2018.4.16-0                           conda-forge --> 2018.03.07-0
    numpy:           1.14.4-py36_blas_openblash24bf2e0_200 conda-forge [blas_openblas] --> 1.13.1-py36_0

Proceed ([y]/n)? n
```
</details>

**Comment by mingwandroid** [comment]
@shoyer this *looks* like a user error. Your second install line has no reference to `conda-forge` so conda knows nothing about it at that point, so it goes ahead and tries to create you a sane environment from what it does know about.

**Comment by dhirschfeld** [comment]
@mingwandroid - toolz depends only on python so the behaviour I'd want when installing toolz is for conda to see the dependency (python) is already installed and *just install toolz* without making any other unnecessary changes.

Whilst conda may not know about `conda-forge` in the 2nd install command, conda knows exactly what's installed in the current environment so should use that information in the solver - i.e. if a valid environment requires that `certifi` be installed conda should see that it is already installed in the current environment, irrespective of whether it knows where it was installed from, and not choose to download and install the exact same version from a known channel.

**Comment by shoyer** [comment]
> this looks like a user error.

Unfortunately, I didn't even make this mistake myself. It's exactly what happens when you use Conda with readthedocs.org:
- read the docs installs your desired environment from a file with `conda env`. In my case, this uses conda-forge.
- read the docs then installs a few unrelated dependencies (e.g., their Sphinx theme) with `conda install` from the default channel.

I'm not sure how the second command could be aware of the channels used to setup the environment, but if that's possible I would gladly fix this for read the docs.

**Comment by mbargull** [comment]
gh-7601 has been merged and should be available once `conda 4.5.9` is released.
An experimental flag `featureless_minimization_disabled_feature_flag` has been added with which the `features` of packages can be completely ignored. This helps with issues like this one that happen on Windows due to packages from `conda-forge` that still have `features: vc`.
Note this **does not** change the behavior around `track_features`, i.e., issues around `conda-forge::numpy` and `conda-forge::blas=*=openblas` still remain until `track_features` is removed from `conda-forge::blas=*=openblas`.

To observe the behavior with that flag enabled for `'zlib <1.2.11' libpng` from @dhirschfeld's example, see:
```
(base) $ conda --version
conda 4.5.8.post17+f41656962
(base) $ export CONDA_SUBDIR=win-64  # convince conda we are on Windows
(base) $ export CONDA_CHANNELS=defaults,conda-forge  # give conda-forge lower priority wrt defaults
(base) $ export CONDA_PINNED_PACKAGES='vc 14.0'  # defaults::vc=14.1 does not have track_features so make sure we do not use that one
(base) $ conda create -qynw 'zlib <1.2.11' libpng
Solving environment: ...working... done

## Package Plan ##

  environment location: /conda/devenv/envs/w

  added / updated specs: 
    - libpng
    - zlib[version='<1.2.11']


The following NEW packages will be INSTALLED:

    libpng:         1.6.30-vc14_1
    vc:             14-h0510ff6_3
    vs2015_runtime: 14.0.25123-3 
    zlib:           1.2.8-vc14_3 

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
(base) $ conda update -qnw --dry-run zlib  # show old/default behavior
Solving environment: ...working... done

## Package Plan ##

  environment location: /conda/devenv/envs/w

  added / updated specs: 
    - zlib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libpng-1.6.34              |           vc14_0         547 KB  conda-forge
    zlib-1.2.11                |       h8395fce_2         237 KB
    ------------------------------------------------------------
                                           Total:         784 KB

The following packages will be UPDATED:

    libpng: 1.6.30-vc14_1 --> 1.6.34-vc14_0     conda-forge [vc14]
    zlib:   1.2.8-vc14_3  --> 1.2.11-h8395fce_2            


DryRunExit: Dry run exiting

(base) $ export CONDA_FEATURELESS_MINIMIZATION_DISABLED_FEATURE_FLAG=1  # enable new behavior
(base) $ conda update -qnw --dry-run zlib
Solving environment: ...working... done

## Package Plan ##

  environment location: /conda/devenv/envs/w

  added / updated specs: 
    - zlib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    zlib-1.2.11                |       h8395fce_2         237 KB

The following packages will be UPDATED:

    zlib: 1.2.8-vc14_3 --> 1.2.11-h8395fce_2


DryRunExit: Dry run exiting
```
